### ISSUE TITLE
Possible false-negative: Removing a manual trait implementation should be a breaking change

### ISSUE THREAD

**Issue by nbigaouette** [obi1kenobi/cargo-semver-checks#990]
### Is this about an existing lint, or proposing a new one?

Deleting the manual implementation of a trait for a type does not trigger a sember-checks error.

### Known issues that might be causing this

- [X] Is the item that should have been flagged originating from another crate, or from a re-export of such an item? ([#638](https://github.com/obi1kenobi/cargo-semver-checks/issues/638))
- [X] Is the issue related to a change in the type of a field, function parameter, generic, or trait bound? ([#149](https://github.com/obi1kenobi/cargo-semver-checks/issues/149))
- [X] If you're proposing a new lint, is it already part of our lint wishlist? ([#5](https://github.com/obi1kenobi/cargo-semver-checks/issues/5))

### Steps to reproduce the bug with the above code

Going from:
```rust
pub trait MyTrait {
    fn method(&self);
}

pub struct MyStruct {}

impl MyTrait for MyStruct {
    fn method(&self) {}
}
```

to:
```rust
pub trait MyTrait {
    fn method(&self);
}

pub struct MyStruct {}
```
is not considered a breaking change by cargo-semver-checks right now. I _think_ it should be?

### Actual Behaviour

`cargo-semver-checks` accepts the removal as non-breaking change.

### Expected Behaviour

`cargo-semver-checks` reporting a semver breaking change error.

### Generated System Information

```
#### Software version

cargo-semver-checks 0.31.0

#### Operating system

macOS 14.7 (Darwin 23.6.0)

#### Command-line

```bash
/Users/nbigaouette/.cargo/bin/cargo-semver-checks semver-checks --bugreport
```

#### cargo version

```
> cargo -V
cargo 1.82.0 (8f40fc59f 2024-08-21)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-apple-darwin
- Family: unix
- OS: macos
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: cmpxchg16b,fxsr,sse,sse2,sse3,sse4.1,ssse3
- Host: x86_64-apple-darwin
```

### Build Configuration

_No response_

### Additional Context

_No response_

**Comment by nbigaouette** [comment]
I searched for other lints that are supposed to check this but couldn't find any. I did start to add a new lint. Not sure how to build the `.ron` file though.

**Comment by obi1kenobi** [comment]
Thanks for opening this and looking into writing a new lint! This is definitely breaking, nice catch :+1:

I'm currently travelling and largely AFK until later this week, but I'd recommend two things:
- Check out the CONTRIBUTING doc's section on adding new lints: https://github.com/obi1kenobi/cargo-semver-checks/blob/main/CONTRIBUTING.md
- Join our [Discord server](https://discord.gg/s5pPdKDh) where many other folks with lint-writing experience will be able to point you in the right direction.

It might also be a good idea to get used to lint-writing by first writing an easier lint that is very similar to another already-implemented lint. That way you can do a side-by-side comparison with an already-merged PR that adds a new lint, and can focus on just tweaking the lint logic and test cases to fit the new case. For example, #946 has one case already implemented and merged, and the remaining cases are _very_ similar.

After you're familiarized with the lint-writing process that way, you could circle back to this lint and implement it more easily.

**Comment by tpoliaw** [comment]
Possibly related to this, keeping the trait impl but tightening the bounds (so that it is no longer implemented for some types).

eg from 

```rust
impl<T> From<T> for Foo {...}
````
to
```rust
impl<T: 'static> From<T> for Foo {...}
```

This caused a breaking change in https://github.com/async-graphql/async-graphql/issues/1634 (not my project) that I can't get to be flagged by cargo-semver-checks

**Comment by obi1kenobi** [comment]
Neat, thank you for the repro and for sharing a real-world case where it happened.

With the latest schema updates as of a week or two ago, I think it's possible to write a lint for "adding new lifetime bound to a generic parameter."

Might you be interested in trying your hand at lint-writing @tpoliaw? I'd be happy to point you in the right direction if so — the time commitment including onboarding time is ~1hr. You're also welcome to join our community Discord if you'd like: https://discord.gg/ywEqqAwq

**Comment by tpoliaw** [comment]
I'm happy to have a look at it but probably won't be for a few days. Will join the discord.

**Comment by obi1kenobi** [comment]
Awesome! The [CONTRIBUTING doc](https://github.com/obi1kenobi/cargo-semver-checks/blob/main/CONTRIBUTING.md#making-your-first-contribution) should have the steps to get started, and feel free to ask questions in the Discord -- lots of people who have already written lints are there and we'd all be happy to help 🙌
